File: src/yanzm/example/adaptersamples/ExtendsArrayAdapterSample.java
package yanzm.example.adaptersamples;

import java.util.ArrayList;
import java.util.List;

import android.app.ListActivity;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

public class ExtendsArrayAdapterSample extends ListActivity {

	public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        List<bindData> data = new ArrayList<bindData>();
        data.add(new bindData("Clock", R.drawable.alermclock));
        data.add(new bindData("Camera", R.drawable.camera));
        
        MyAdapter adapter = new MyAdapter(this, R.layout.list_item5, data);
        setListAdapter(adapter);
    }
	
	class bindData {
		String text;
		int imageResourceId;
		
		public bindData(String text, int imageResourceId) {
			this.text = text;
			this.imageResourceId = imageResourceId;
		}
	}

	static class ViewHolder {
		TextView textView;
		ImageView imageView;
	} 

	public class MyAdapter extends ArrayAdapter<bindData> {
		private LayoutInflater inflater;
		private int layoutId;

		public MyAdapter(Context context, int layoutId, List<bindData> objects) {
			super(context, 0, objects);
			this.inflater = (LayoutInflater) context
					.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
			this.layoutId = layoutId;
		}

		@Override
		public View getView(int position, View convertView, ViewGroup parent) {
			ViewHolder holder;

			if (convertView == null) {
				convertView = inflater.inflate(layoutId, parent, false);
				holder = new ViewHolder();
				holder.textView = (TextView) convertView.findViewById(R.id.textview);
				holder.imageView = (ImageView) convertView.findViewById(R.id.imageview);
				convertView.setTag(holder);
			} else {
				holder = (ViewHolder) convertView.getTag();
			}
			bindData data= getItem(position);
			holder.textView.setText(data.text);
			holder.imageView.setImageResource(data.imageResourceId);
			return convertView;
		}
	}
}

File: app/src/main/java/yanzm/example/adaptersamples/ExtendsArrayAdapterSample.kt
package yanzm.example.adaptersamples

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class ExtendsArrayAdapterSample : AppCompatActivity() {

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val listView = ListView(this)
        setContentView(listView)

        val data = listOf(
            Item("Clock", R.drawable.ic_alarm),
            Item("Camera", R.drawable.ic_camera),
        )

        listView.adapter = MyAdapter(this, R.layout.list_item5, data)
    }

    data class Item(val text: String, val imageResourceId: Int)

    data class ViewHolder(val itemView: View) {
        private val textView: TextView = itemView.findViewById(R.id.textview)
        private val imageView: ImageView = itemView.findViewById(R.id.imageview)

        fun bind(item: Item) {
            textView.text = item.text
            imageView.setImageResource(item.imageResourceId)
        }
    }

    class MyAdapter(context: Context, private val layoutId: Int, objects: List<Item>) :
        ArrayAdapter<Item?>(context, 0, objects) {

        private val inflater = LayoutInflater.from(context)

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val holder: ViewHolder = if (convertView == null) {
                val view = inflater.inflate(layoutId, parent, false)
                ViewHolder(view).apply {
                    view.tag = this
                }
            } else {
                convertView.tag as ViewHolder
            }

            val item = getItem(position)!!
            holder.bind(item)

            return holder.itemView
        }
    }
}
